File: orchestrator/routes/api_v1/documents.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Lumogis
"""Document library API — list, detail, delete, re-ingest (LUM-160)."""

from __future__ import annotations

from auth import UserContext
from authz import require_user
from fastapi import APIRouter
from fastapi import Depends
from fastapi import HTTPException
from fastapi import status
from models.api_v1 import DocumentDeleteResponse
from models.api_v1 import DocumentDetail
from models.api_v1 import DocumentListResponse
from models.api_v1 import ReingestQueuedResponse
from models.api_v1 import ReingestRequest
from models.api_v1 import ShareQueuedResponse
from services.document_purge import DocumentNotFoundError
from services.documents import DocumentNotSharedError
from services.documents import SourceUnavailableError

from fastapi import Body

from services import documents as doc_svc

router = APIRouter(
    prefix="/api/v1/documents",
    tags=["v1-documents"],
    dependencies=[Depends(require_user)],
)


def _parse_document_id(document_id: str) -> int:
    try:
        parsed = int(document_id)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"error": "invalid_document_id"},
        ) from exc
    if parsed <= 0:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"error": "invalid_document_id"},
        )
    return parsed


def _parse_limit(limit: int) -> int:
    if limit <= 0 or limit > 100:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"error": "invalid_limit"},
        )
    return limit


@router.get("", response_model=DocumentListResponse)
def list_documents(
    user: UserContext = Depends(require_user),
    limit: int = 50,
) -> DocumentListResponse:
    lim = _parse_limit(limit)
    rows = doc_svc.list_documents(user, limit=lim)
    return DocumentListResponse(documents=rows)


@router.get("/{document_id}", response_model=DocumentDetail)
def get_document(
    document_id: str,
    user: UserContext = Depends(require_user),
) -> DocumentDetail:
    doc_id = _parse_document_id(document_id)
    try:
        return doc_svc.get_document(user, doc_id)
    except DocumentNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": "document_not_found"},
        ) from exc


@router.delete("/{document_id}", response_model=DocumentDeleteResponse)
def delete_document(
    document_id: str,
    user: UserContext = Depends(require_user),
) -> DocumentDeleteResponse:
    doc_id = _parse_document_id(document_id)
    try:
        result = doc_svc.delete_document(user.user_id, doc_id)
    except DocumentNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": "document_not_found"},
        ) from exc
    return DocumentDeleteResponse(
        document_id=doc_id,
        deleted=result.postgres_deleted,
        partial=result.partial,
        errors=result.errors,
    )


@router.post(
    "/{document_id}/reingest",
    status_code=status.HTTP_202_ACCEPTED,
    response_model=ReingestQueuedResponse,
)
def reingest_document(
    document_id: str,
    body: ReingestRequest | None = None,
    user: UserContext = Depends(require_user),
) -> ReingestQueuedResponse:
    doc_id = _parse_document_id(document_id)
    force = body.force if body is not None else False
    try:
        return doc_svc.reingest_document(user.user_id, doc_id, force=force)
    except DocumentNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": "document_not_found"},
        ) from exc
    except SourceUnavailableError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={"error": "source_unavailable"},
        ) from exc


@router.post(
    "/{document_id}/publish",
    status_code=status.HTTP_202_ACCEPTED,
    response_model=ShareQueuedResponse,
)
def publish_document(
    document_id: str,
    body: dict = Body(default={}),
    user: UserContext = Depends(require_user),
) -> ShareQueuedResponse:
    """Share a personal document with the household (LUM-157).

    Owner-only, ``scope='personal'`` source (validated synchronously → 404 on a
    foreign/non-personal document; 400 on an invalid scope), then enqueues the
    background ``share_document`` projection job (→ 202 + job_id).
    """
    doc_id = _parse_document_id(document_id)
    scope = (body or {}).get("scope", "shared")
    if scope != "shared":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"error": "invalid_scope"},
        )
    try:
        return doc_svc.share_document(user.user_id, doc_id)
    except DocumentNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": "document_not_found"},
        ) from exc


@router.delete(
    "/{document_id}/publish",
    status_code=status.HTTP_202_ACCEPTED,
    response_model=ShareQueuedResponse,
)
def unpublish_document(
    document_id: str,
    user: UserContext = Depends(require_user),
) -> ShareQueuedResponse:
    """Stop sharing a document with the household (LUM-157)."""
    doc_id = _parse_document_id(document_id)
    try:
        return doc_svc.unshare_document(user.user_id, doc_id)
    except (DocumentNotFoundError, DocumentNotSharedError) as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": "document_not_found"},
        ) from exc

File: clients/lumogis-web/src/api/documents.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Lumogis
//
// Document library API client (LUM-160).

import type { ApiClient } from "./client";

export type DocumentStatus = "indexing" | "indexed" | "failed";

// LUM-157 — household document sharing lifecycle. ``partial`` is a
// terminal-but-incomplete share (some sections still indexing).
export type ShareStatus =
  | "personal"
  | "sharing"
  | "shared"
  | "unsharing"
  | "partial";

export interface DocumentSummary {
  document_id: number | null;
  in_flight_job_id?: number | null;
  display_name: string;
  file_path: string;
  file_type: string;
  chunk_count: number;
  entity_count: number;
  scope: string;
  status: DocumentStatus;
  indexed_at: string | null;
  error_message: string | null;
  // LUM-157 — additive + defaulted server-side, so older payloads still parse.
  share_status?: ShareStatus;
  in_flight_share_job_id?: number | null;
  is_owner?: boolean;
  // LUM-585 — "Shared by {member}" attribution (non-owner detail view only).
  shared_by?: string | null;
}

export interface DocumentEntityLink {
  entity_id: string;
  name: string;
  entity_type: string;
}

export interface DocumentDetail extends DocumentSummary {
  file_hash: string | null;
  entities: DocumentEntityLink[];
  source_available: boolean;
}

export interface DocumentListResponse {
  documents: DocumentSummary[];
}

export interface DocumentDeleteResponse {
  document_id: number;
  deleted: boolean;
  partial: boolean;
  errors: string[];
}

export interface ReingestQueuedResponse {
  document_id: number;
  job_id: number;
  queued: boolean;
}

export interface ShareQueuedResponse {
  document_id: number;
  job_id: number;
  share_status: "sharing" | "unsharing";
}

/** True when the share lifecycle is in a state the badge should treat as shared. */
export function isShared(status: ShareStatus | undefined): boolean {
  return status === "shared" || status === "partial";
}

export async function listDocuments(
  client: ApiClient,
  signal?: AbortSignal,
): Promise<DocumentListResponse> {
  return client.getJson<DocumentListResponse>("/api/v1/documents", { signal });
}

export async function getDocument(
  client: ApiClient,
  documentId: number,
  signal?: AbortSignal,
): Promise<DocumentDetail> {
  return client.getJson<DocumentDetail>(
    `/api/v1/documents/${encodeURIComponent(String(documentId))}`,
    { signal },
  );
}

export async function deleteDocument(
  client: ApiClient,
  documentId: number,
): Promise<DocumentDeleteResponse> {
  return client.delete<DocumentDeleteResponse>(
    `/api/v1/documents/${encodeURIComponent(String(documentId))}`,
  );
}

export async function reingestDocument(
  client: ApiClient,
  documentId: number,
  body?: { force?: boolean },
): Promise<ReingestQueuedResponse> {
  return client.postJson<{ force?: boolean }, ReingestQueuedResponse>(
    `/api/v1/documents/${encodeURIComponent(String(documentId))}/reingest`,
    body ?? {},
  );
}

/** Share a personal document with the household (LUM-157) → 202 + job_id. */
export async function publishDocument(
  client: ApiClient,
  documentId: number,
): Promise<ShareQueuedResponse> {
  return client.postJson<Record<string, never>, ShareQueuedResponse>(
    `/api/v1/documents/${encodeURIComponent(String(documentId))}/publish`,
    {},
  );
}

/** Stop sharing a document with the household (LUM-157) → 202 + job_id. */
export async function unpublishDocument(
  client: ApiClient,
  documentId: number,
): Promise<ShareQueuedResponse> {
  return client.delete<ShareQueuedResponse>(
    `/api/v1/documents/${encodeURIComponent(String(documentId))}/publish`,
  );
}
